src/osd/scrubber/scrub_backend.h

```diff
@@ -0,0 +1,419 @@
+#include <fmt/format.h>
+
+#include <string_view>
+
+#include "common/LogClient.h"
+
+class PG;
+class PgScrubber;
+class PGBackend;
+
+
+using data_omap_digests_t =
+  std::pair<std::optional<uint32_t>, std::optional<uint32_t>>;
+
+using shard_info_map_t = std::map<pg_shard_t, shard_info_wrapper>;
+
+using IterToSMap = std::map<pg_shard_t, ScrubMap>::iterator;
+
+/**
+ * A structure used internally by select_auth_object()
+ *
+ * Conveys the usability of a specific shard as an auth source.
+ */
+struct shard_as_auth_t {
+  enum class usable_t : uint8_t { not_usable, usable };
+
+  // the ctor used when the shard should not be considered as auth
+  explicit shard_as_auth_t(std::string err_msg)
+      : possible_auth{usable_t::not_usable}
+      , error_text{err_msg}
+      , oi{}
+      , auth_iter{}
+      , digest{std::nullopt}
+  {}
+
+  shard_as_auth_t(std::string err_msg, std::optional<uint32_t> data_digest)
+      : possible_auth{usable_t::not_usable}
+      , error_text{err_msg}
+      , oi{}
+      , auth_iter{}
+      , digest{data_digest}
+  {}
+
+  // possible auth candidate
+  shard_as_auth_t(const object_info_t& anoi,
+                  IterToSMap it,
+                  std::string err_msg,
+                  std::optional<uint32_t> data_digest)
+      : possible_auth{usable_t::usable}
+      , error_text{err_msg}
+      , oi{anoi}
+      , auth_iter{it}
+      , digest{data_digest}
+  {}
+
+
+  usable_t possible_auth;
+  std::string error_text;
+  object_info_t oi;
+  IterToSMap auth_iter;
+  std::optional<uint32_t> digest;
+  // when used for Crimson, we'll probably want to return 'digest_match' (and
+  // other in/out arguments) via this struct
+};
+
+struct auth_selection_t {
+  IterToSMap auth;        ///< an iter into one of this_chunk->maps
+  pg_shard_t auth_shard;  // set to auth->first
+  object_info_t auth_oi;
+  shard_info_map_t shard_map;
+  bool is_auth_available{false};  // managed to select an auth' source?
+  bool digest_match{true};        // do all (existing) digests match?
+};
+
+// note: some scrub tests are sensitive to the specific format of
+// auth_selection_t listing in the logs
+template <>
+struct fmt::formatter<auth_selection_t> {
+  template <typename ParseContext>
+  constexpr auto parse(ParseContext& ctx)
+  {
+    return ctx.begin();
+  }
+
+  template <typename FormatContext>
+  auto format(auth_selection_t const& aus, FormatContext& ctx)
+  {
+    return fmt::format_to(ctx.out(),
+                          " {{AU-S: {}->{:x} OI({:x}:{}) {} dm:{}}} ",
+                          aus.auth->first,
+                          (uint64_t)(&aus.auth->second),
+                          (uint64_t)(&aus.auth_oi),
+                          aus.auth_oi,
+                          aus.shard_map.size(),
+                          aus.digest_match);
+  }
+};
+
+/**
+ * the back-end data that is per-chunk
+ *
+ * Created by the Scrubber after all replicas' maps have arrived.
+ */
+struct ScrubBeChunk {
```

Comment:
scrub_chunk_t to match other names in this file.

Reply:
ScrubBackend::new_chunk() exposes this entire structure to PGScrubber.  I'm not sure what we're trying to encapsulate here.

Reply:
> scrub_chunk_t to match other names in this file.

Followed the guidance that very simple constructs should be named with a '_t', while the
rest using Camel. Can change if that's the pref.